src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -923,6 +964,767 @@ private FindRequestContext ConstructFindRequestContext(RestRequestContext contex
+                                                      statusCode: HttpStatusCode.InternalServerError,
+                                                      subStatusCode: DataApiBuilderException.SubStatusCodes.DatabaseOperationFailed);
+                }
+                else
+                {
+                    if (dbResultSetRow is null)
+                    {
+                        throw new DataApiBuilderException(message: "No data returned back from database.",
+                                                          statusCode: HttpStatusCode.InternalServerError,
+                                                          subStatusCode: DataApiBuilderException.SubStatusCodes.DatabaseOperationFailed);
+                    }
+                    else
+                    {
+                        throw new DataApiBuilderException(message: $"Could not insert row with given values for entity: {entityName} at nesting level : {nestingLevel}",
+                                                          statusCode: HttpStatusCode.Forbidden,
+                                                          subStatusCode: DataApiBuilderException.SubStatusCodes.DatabasePolicyFailure);
+                    }
+                }
+            }
+
+            return dbResultSetRow.Columns;
+        }
+
+        /// <summary>
+        /// Helper method to extract the primary key fields from all the fields of the entity.
+        /// </summary>
+        /// <param name="sqlMetadataProvider">SqlMetadaProvider object for the given database</param>
+        /// <param name="entityName">Name of the entity</param>
+        /// <param name="entityFields">All the fields belonging to the entity</param>
+        /// <returns>Primary Key fields</returns>
+        private static Dictionary<string, object?> FetchPrimaryKeyFieldValues(ISqlMetadataProvider sqlMetadataProvider, string entityName, Dictionary<string, object?> entityFields)
+        {
+            Dictionary<string, object?> pkFields = new();
+            SourceDefinition sourceDefinition = sqlMetadataProvider.GetSourceDefinition(entityName);
+            foreach (string primaryKey in sourceDefinition.PrimaryKey)
+            {
+                if (sqlMetadataProvider.TryGetExposedColumnName(entityName, primaryKey, out string? name)
+                    && entityFields.TryGetValue(name, out object? value)
+                    && value != null)
+                {
+                    pkFields.Add(primaryKey, value);
+                }
+                else
+                {
+                    throw new DataApiBuilderException(message: $"Primary key field {name} has null value but it is expected to have a non-null value",
+                                                      statusCode: HttpStatusCode.InternalServerError,
+                                                      subStatusCode: DataApiBuilderException.SubStatusCodes.UnexpectedError);
```

Comment:
shouldn't this be http400 since user provided null value? `entityFields` seems vague -> entityFields returned by database query results or entityFields that are provided in request?

Reply:
`entityFields` is supposed to contain the values with which a database record was created for the entity. The function tries to extract only the primary key fields as only pk fields are needed for processing the selection set. 

So, when a pk field has a `null` value here, it means something has gone wrong during the creation of the record in the database. Hence, throwing exception with `InternalServerError` status code and `UnexpectedError` sub-status code. 

Agree that `entityFields` is a bit vague. Re-named it to `createdValuesForEntityItem`.